### PR TITLE
APDS-61 - [BE] add API to list and delete permanent tokens

### PR DESCRIPTION
**Pull request by poxip**
This also closes APDS-60. 

In this PR I had to drop support for DRF 3.1 and 3.2, because their code is incompatible with Django >= 1.10 (see: https://github.com/encode/django-rest-framework/issues/3252 - it has been fixed in 3.3). I can't do anything about it to make it backwards compatible.

### CODE REVIEW

File: rest_framework_jwt/views.py [hunk 1]
```diff
@@ -99,6 +101,23 @@ class RefreshJSONWebToken(JSONWebTokenAPIView):
...
 
+class DeviceViewSet(mixins.ListModelMixin, mixins.DestroyModelMixin, viewsets.GenericViewSet):
+    queryset = Device.objects.all()
+    serializer_class = DeviceSerializer
+    permission_classes = [IsAuthenticated]
+
+    def get_queryset(self):
+        return self.queryset.filter(user=self.request.user)
+
+    def get(self, *args, **kwargs):
```
Comment (remik) [inline rest_framework_jwt/views.py]: Why do we need this and below if you have mixins?

File: tests/test_views.py [hunk 2]
```diff
@@ -492,3 +495,38 @@ def test_refresh_jwt_after_refresh_expiration(self):
     def tearDown(self):
         # Restore original settings
         api_settings.JWT_ALLOW_REFRESH = DEFAULTS['JWT_ALLOW_REFRESH']
+
+
+class DeviceViewTests(TokenTestCase):
```
Comment (remik) [inline tests/test_views.py]: Please add devices for not your user to make sure we are not showing it and can't delete.

File: tests/urls.py [hunk 3]
```diff
@@ -28,6 +28,8 @@ def post(self, request):
     url(r'^auth-token/$', views.obtain_jwt_token),
     url(r'^auth-token-refresh/$', views.refresh_jwt_token),
     url(r'^auth-token-verify/$', views.verify_jwt_token),
+    url(r'^devices/(?P<pk>[\w]+)$', views.device_delete_view),
```
Comment (remik) [inline tests/urls.py]: Use router.

**Review by remik — CHANGES_REQUESTED** [review]

File: rest_framework_jwt/views.py (hunk 1 above)
Comment (poxip) [reply rest_framework_jwt/views.py]: Mixins used here only add the "destroy" and "list" methods

Comment (remik) [reply rest_framework_jwt/views.py]: We used mixins so router should discover it. Check: http://www.django-rest-framework.org/api-guide/viewsets/

**Comment by poxip** [comment]
The tests are okay (see details), Travis does not provide Python3.3 somehow (it works in the original repo though). We do not have to drop support for DRF 3.0, 3.1, 3.2 because those versions are only supposed to work with Django 1.8 and 1.9.

File: tests/test_views.py (hunk 2 above)
Comment (poxip) [reply tests/test_views.py]: Added in the CR commit.

File: rest_framework_jwt/models.py [hunk 4]
```diff
@@ -1 +1,16 @@
...
+    """
+    Device model used for permanent token authentication
+    """
+    permanent_token = models.CharField(max_length=255, primary_key=True)
+    jwt_secret = models.CharField(max_length=255)
+    created = models.DateTimeField(auto_now_add=True)
+    user = models.ForeignKey(settings.AUTH_USER_MODEL, on_delete=models.CASCADE)
+    name = models.CharField(_('Device name'), max_length=255)
+    details = models.CharField(_('Device details'), max_length=255, blank=True)
+    last_request_datetime = models.DateTimeField(auto_now=True)
```
Comment (jacoor) [inline rest_framework_jwt/models.py:20]: Missing empty line at the end of file.
Pylint and pep8 will complain.

File: rest_framework_jwt/serializers.py [hunk 5]
```diff
@@ -169,3 +170,9 @@ def validate(self, attrs):
             'token': jwt_encode_handler(new_payload),
             'user': user
         }
+
+
+class DeviceSerializer(serializers.ModelSerializer):
+    class Meta:
+        model = Device
+        fields = ['permanent_token', 'jwt_secret', 'created', 'name', 'details', 'last_request_datetime']
```
Comment (jacoor) [inline rest_framework_jwt/serializers.py]: WHat?! you return JWT secret to the world? 
http://getblimp.github.io/django-rest-framework-jwt/#jwt_secret_key Please check documentation what JWT secret is.

File: tests/test_views.py [hunk 6]
```diff
@@ -492,3 +493,46 @@ def test_refresh_jwt_after_refresh_expiration(self):
...
+        self.assertEqual(Device.objects.filter(permanent_token=self.device.permanent_token).count(), 0)
+
+    def test_device_list(self):
+        client = APIClient(enforce_csrf_checks=True)
+        self._login(client)
+        response = client.get('/devices/', format='json')
+        self.assertEqual(response.status_code, 200)
+        self.assertEqual(len(response.data), 1)
+        self.assertEqual(set(response.data[0].keys()), {
+            'permanent_token', 'jwt_secret', 'created', 'name', 'details', 'last_request_datetime'
```
Comment (jacoor) [inline tests/test_views.py]: Please, do not make any field *_secret_* public. After all, there is a reason why they call it secret.
http://getblimp.github.io/django-rest-framework-jwt/#jwt_secret_key

**Review by jacoor — CHANGES_REQUESTED** [review]

File: rest_framework_jwt/models.py (hunk 4 above)
Comment (poxip) [reply rest_framework_jwt/models.py:20]: GitHub does not show the last empty line as far as I remember. The flake8 task in tox did not complain, see: https://travis-ci.org/ArabellaTech/django-rest-framework-jwt/jobs/256993179

**Review by remik — APPROVED** [review]

**Pull request merged**